packages/server-wallet/src/onchain-service/transaction-submission.ts

```diff
@@ -0,0 +1,120 @@
+import {providers, Wallet} from 'ethers';
+import PriorityQueue from 'p-queue';
+
+import {
+  TransactionSubmissionServiceInterface,
+  MinimalTransaction,
+  TransactionSubmissionOptions,
+} from './types';
+
+// FIXME: replace with
+// import {WalletError as ChannelWalletError} from '@statechannels/server-wallet';
+import {WalletError as ChannelWalletError} from '..';
+
+type Values<E> = E[keyof E];
+
+export class TransactionSubmissionError extends ChannelWalletError {
+  readonly type = ChannelWalletError.errors.OnchainError;
+
+  // Errors where transactions will be safely retried
+  static readonly knownErrors = {
+    badNonce: `the tx doesn't have the correct nonce`,
+    invalidNonce: 'Invalid nonce',
+    noHash: 'no transaction hash found in tx response',
+    underpricedReplacement: 'replacement transaction underpriced',
+  };
+
+  // Reasons an error is thrown from the transaction submission
+  // service
+  static readonly reasons = {
+    zeroAttempts: 'Invalid max transaction submission attempt count of 0',
+    failedAllAttempts: 'Failed all transaction attempts',
+    unknownError: 'Transaction failed with unkown error',
+  } as const;
+  constructor(
+    reason: Values<typeof TransactionSubmissionError.reasons>,
+    public readonly data: any = undefined
+  ) {
+    super(reason);
+  }
+}
+
+export class TransactionSubmissionService implements TransactionSubmissionServiceInterface {
+  // Chain provider
+  private provider: providers.JsonRpcProvider;
+
+  // Wallet to send transactions
+  private wallet: Wallet;
+
+  // Queues to handle transactions
+  private queue = new PriorityQueue({concurrency: 1});
+
+  // Memory tracking wallet nonce to assist with retries
+  // TODO: maybe replace with NonceManager?
+  private memoryNonce = 0;
+
+  constructor(provider: string | providers.JsonRpcProvider, wallet: string | Wallet) {
+    this.provider =
+      typeof provider === 'string' ? new providers.JsonRpcProvider(provider) : provider;
+    this.wallet =
+      typeof wallet === 'string'
+        ? new Wallet(wallet, this.provider)
+        : wallet.connect(this.provider);
+  }
+
+  public async submitTransaction(
+    tx: MinimalTransaction,
+    options: TransactionSubmissionOptions = {}
+  ): Promise<providers.TransactionResponse> {
+    // Check via casting to avoid 0 returning falsy values
+    const attempts = typeof options.maxSendAttempts === 'number' ? options.maxSendAttempts : 1;
```

Comment:
Since we're using typescript, we could use the [null coalescing operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator). (The compiled javascript will be somewhat similar to this code)

```suggestion
    const attempts = options.maxSendAttempts ?? 1;
```